### PR TITLE
Disambiguate alter commands list

### PR DESCRIPTION
**Pull request by antaljanosbenjamin**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Allow alter operations to be surrounded by parenthesis. The emission of parentheses can be controlled by the `format_alter_operations_with_parentheses` config. By default in formatted queries the parentheses are emitted as we store the formatted alter operations in some places as metadata (e.g.: mutations). The new syntax clarifies some of the queries where alter operations end in a list. E.g.: `ALTER TABLE x MODIFY TTL date GROUP BY a, b, DROP COLUMN c` cannot be parsed properly with the old syntax. In the new syntax the query `ALTER TABLE x (MODIFY TTL date GROUP BY a, b), (DROP COLUMN c)` is obvious. Older versions are not able to read the new syntax, therefore using the new syntax might cause issues if newer and older version of ClickHouse are mixed in a single cluster.

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

### CODE REVIEW

**Comment by robot-clickhouse-ci-1** [comment]
*This is an automated comment for commit 3049afd5c98916e068c785822c7130e33da15632 with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/59532/3049afd5c98916e068c785822c7130e33da15632/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td><a href="https://github.com/ClickHouse/ClickHouse/blob/master/.github/PULL_REQUEST_TEMPLATE.md?plain=1">❌ failure</a></td></tr>
<tbody>
</table>

**Comment by UnamedRus** [comment]
Can it cover as well https://github.com/ClickHouse/ClickHouse/issues/17303?

**Comment by antaljanosbenjamin** [comment]
@UnamedRus I will check that issue, but it seems like not a parsing, but a logical issue there.

**Comment by UnamedRus** [comment]
Actually yeah, my memory fails me.
There was another issue with comma, but for TTL GROUP BY  https://github.com/ClickHouse/ClickHouse/issues/17304 and it fixed.

**Comment by antaljanosbenjamin** [comment]
The scope of the PR is changed: because the necessary part is only the alter commands and the TTL elements are harder to parse properly (`(expr)` can be either a TTL element with parens inside and with the expression `expr`, or just a TTL element without parens with the expression `(expr)`, it won't try to fix that now.

**Comment by antaljanosbenjamin** [comment]
> it fixed.

Actually there is still an issue if you don't use the `SET` keyword, so e.g.:
```
CREATE TABLE test_table
(
    `key_a` UInt32,
    `key_b` UInt32,
    `ts` DateTime,
    `value` UInt32
)
ENGINE = MergeTree
PARTITION BY tuple()
ORDER BY (key_a, key_b)
TTL
    ts GROUP BY key_a, key_b,
    ts + toIntervalMonth(2)


Received exception from server (version 24.2.1):
Code: 450. DB::Exception: Received from localhost:9000. DB::Exception: TTL Expression GROUP BY key should be a prefix of primary key. (BAD_TTL_EXPRESSION)
```

Here `ts + toIntervalMonth(2)` is parsed as `GROUP BY`, which is bad. I will create an issue about this.

**Comment by UnamedRus** [comment]
BTW, when change happened to D DDL worker, it was done via user level setting `distributed_ddl_entry_format_version`, with ~1 year delay

https://github.com/ClickHouse/ClickHouse/pull/35463

File: src/Parsers/ASTTTLElement.cpp [hunk 1]
```diff
@@ -1,8 +1,10 @@
-#include <Columns/Collator.h>
-#include <Common/quoteString.h>
 #include <Parsers/ASTTTLElement.h>
+
+#include <Columns/Collator.h>
+#include <Core/ServerSettings.h>
```
Comment (antaljanosbenjamin) [inline src/Parsers/ASTTTLElement.cpp:4]: revert these changes.

File: src/Parsers/ExpressionElementParsers.cpp [hunk 2]
```diff
@@ -2224,6 +2224,9 @@ bool ParserTTLElement::parseImpl(Pos & pos, ASTPtr & node, Expected & expected)
     ParserKeyword s_remove("REMOVE");
     ParserKeyword s_modify("MODIFY");
 
+    ParserToken parser_opening_round_bracket(TokenType::OpeningRoundBracket);
```
Comment (antaljanosbenjamin) [inline src/Parsers/ExpressionElementParsers.cpp:2227]: revert

File: src/Parsers/ASTAlterQuery.h [hunk 3]
```diff
@@ -222,10 +222,16 @@ class ASTAlterCommand : public IAST
 
     ASTPtr clone() const override;
 
+    // This function is only meant to be called during application startup
+    // For reasons see https://github.com/ClickHouse/ClickHouse/pull/59532
+    static void setFormatAlterCommandsWithParentheses(bool value) { format_alter_commands_with_parentheses = value; }
```
Comment (tavplubix) [inline src/Parsers/ASTAlterQuery.h:227]: Should we call it in `clickhouse-client`  and `clickhouse-local` as well? They print a formatted query:

```
localhost.localdomain :) SeLeCt NuLl

SELECT NULL

Query id: fdbaa7b8-75c0-4f58-b79c-46b9ae290e1e

┌─NULL─┐
│ ᴺᵁᴸᴸ │
└──────┘

1 row in set. Elapsed: 0.005 sec. 
```

I think it's fine to unconditionally set it to `true` for that purpose

File: src/Interpreters/InterpreterKillQueryQuery.cpp [hunk 4]
```diff
@@ -277,9 +277,11 @@ BlockIO InterpreterKillQueryQuery::execute()
                     code = CancellationCode::NotFound;
                 else
                 {
-                    ParserAlterCommand parser;
+                    const auto alter_command = command_col.getDataAt(i).toString();
+                    const auto with_round_bracket = alter_command.front() == '(';
+                    ParserAlterCommand parser{with_round_bracket};
```
Comment (tavplubix) [inline src/Interpreters/InterpreterKillQueryQuery.cpp:282]: Does it make sense to always allow round brackets when parsing, despite the value of the setting? So the setting will be only used for formatting, and the parser will be able to always understand both versions

File: tests/integration/test_unambiguous_alter_commands/configs/format_alter_operations_with_parentheses.xml [hunk 5]
```diff
@@ -0,0 +1,3 @@
+<clickhouse>
+    <format_alter_operations_with_parentheses>1</format_alter_operations_with_parentheses>
```
Comment (tavplubix) [inline tests/integration/test_unambiguous_alter_commands/configs/format_alter_operations_with_parentheses.xml:2]: We can enable it in functional tests by adding this config to `tests/config/install.sh`
Btw, functional tests are much better than integration tests: https://github.com/ClickHouse/ClickHouse/pull/39359#issuecomment-1188296170
(but it's not necessary to rewrite existing tests)

File: src/Interpreters/InterpreterKillQueryQuery.cpp (hunk 4 above)
Comment (antaljanosbenjamin) [reply src/Interpreters/InterpreterKillQueryQuery.cpp:282]: The idea is by allowing parsing with parentheses by default is the transition can be smoother during migrating to newer versions. E.g. during live migration if there are one replica which produces alter commands with parens, then the older replicas will parse it properly.

File: tests/integration/test_unambiguous_alter_commands/configs/format_alter_operations_with_parentheses.xml (hunk 5 above)
Comment (antaljanosbenjamin) [reply tests/integration/test_unambiguous_alter_commands/configs/format_alter_operations_with_parentheses.xml:2]: I checked this, and if we would change this in functional tests, then we would change this for all of the functional tests. I think it shouldn't cause an issue, but we would test something that is not the "default" behavior of clickhouse.

Furthermore I used the same queries in functional tests, but expecting the old behavior. So I would say the only question is which way should be the default in functional tests? The other one has to be tested in integration test either way, because we have to change the server setting.

File: src/Interpreters/InterpreterKillQueryQuery.cpp (hunk 4 above)
Comment (antaljanosbenjamin) [reply src/Interpreters/InterpreterKillQueryQuery.cpp:282]: Maybe my previous answer wasn't the best. So I wanted to avoid allowing the "mix" of the old and new syntax. So in an alter command list each command must or mustn't have to be surrounded with parentheses. E.g.: `ALTER TABLE x MODIFY TTL, DROP COLUM a` is fine, `ALTER TABLE x (MODIFY TTL), (DROP COLUMN a)` also fine, but `ALTER TABLE x (MODIFY TTL), DROP COLUMN a` is not fine.

This means in some way `ParserAlterCommandList` has to be able to enforce it. This could be done in two ways in my opinion:
1. The list parser figures this out and passes it to the individual command parser.
2. The individual command parser decides whether parens should be used or not and because in `ParserAlterCommandList::parseImpl` only a single parser is used, it can carry over this information. It is doable, but it would introduce in my opnion a hidden relation between the first parsed command and every command after that. I think right now parsers doesn't have this kind of "memory" in general. Thus, I opted for option 1.

That's why here determining `with_round_bracket` is necessary. What do you think?

Comment (tavplubix) [reply src/Interpreters/InterpreterKillQueryQuery.cpp:282]: > So in an alter command list each command must or mustn't have to be surrounded with parentheses.

Yes, I agree.

> This means in some way ParserAlterCommandList has to be able to enforce it. ... That's why here determining with_round_bracket is necessary. 

Sorry, I missed this when reading the code, it was unobvious.
Yes, the first option is better because in the worst case, someone will pass `true` or `false` unconditionally instead of proper `with_round_bracket` flag (so only one version of the syntax will work in this place). And with the second option, someone may use one `ParserAlterCommand` for multiple queries, and it will cause an obscure bug

**Review by tavplubix — APPROVED** [review]

**Comment by antaljanosbenjamin** [comment]
Test failures:
- `AST Fuzzer (ubsan)` failed with `Logical error: 'Expected single dictionary argument for function.'.` #58979 
- `Integration Tests (asan, Analyzer)`, `Integration Tests (tsan)` failed `test_replicated_database/test.py::test_startup_without_zk` #60215
- `Stateless Tests (tsan, S3 Storage)` failed `00984_parser_stack_overflow` #60216
- `Stress Test (msan)` failed with a fatal error, see below.

```
2024.02.21 02:22:36.637567 [ 33028 ] {} <Trace> DatabaseAtomic (test_15): Loading table test_15.test_param_view2
2024.02.21 02:22:36.637759 [ 33028 ] {} <Fatal> : Logical error: 'Invalid storage definition in metadata file: it's a bug or result of manual intervention in metadata files'.
```

The table `test_param_view2` is created in `tests/queries/0_stateless/02818_parameterized_view_with_cte_multiple_usage.sql`:
```
create view test_param_view2 as
with {param_test_val:UInt8} as param_test_val
select param_test_val,
       arrayCount((a)->(a < param_test_val), t.arr) as cnt1,
       arrayCount((a)->(a < param_test_val+1), t.arr) as cnt2
from (select [1,2,3,4,5] as arr) t;
```

I don't think it has anything to do with my changes, so I will merge.

**Comment by antaljanosbenjamin** [comment]
- `test_backup_restore_s3/test.py::test_user_specific_auth` https://github.com/ClickHouse/ClickHouse/pull/60144#discussion_r1496391457
- `test_materialized_mysql_database/test.py::test_table_table` https://github.com/ClickHouse/ClickHouse/pull/59082
- `01747_system_session_log_long`, `01702_system_query_log`, `01656_test_query_log_factories_info`, `01600_log_queries_with_extensive_info` https://github.com/ClickHouse/ClickHouse/issues/57984
